scene/3d/node_3d.cpp

```diff
@@ -341,6 +349,112 @@ Transform3D Node3D::get_transform() const {
+	// (For instance, get_global_transform_interpolated() could be called multiple times.)
+	if (pid.current_physics_tick != tick) {
+		// Timeout?
+		if (tick >= pid.timeout_physics_tick) {
+			return false;
+		}
+
+		if (pid.current_physics_tick == (tick - 1)) {
+			// Normal interpolation situation, there is a continuous flow of data
+			// from one tick to the next...
+			pid.global_xform_prev = pid.global_xform_curr;
+		} else {
+			// There has been a gap, we cannot sensibly offer interpolation over
+			// a multitick gap, so we will teleport.
+			pid.global_xform_prev = get_global_transform();
+		}
+		pid.current_physics_tick = tick;
+	}
+
+	pid.global_xform_curr = get_global_transform();
+	return true;
+}
+
+void Node3D::_disable_client_physics_interpolation() {
+	// Disable any current client side interpolation.
+	// (This can always restart as normal if you later re-attach the node to the SceneTree.)
+	if (data.client_physics_interpolation_data) {
+		memdelete(data.client_physics_interpolation_data);
+		data.client_physics_interpolation_data = nullptr;
+
+		SceneTree *tree = get_tree();
+		if (tree && _client_physics_interpolation_node_3d_list.in_list()) {
+			tree->client_physics_interpolation_remove_node_3d(&_client_physics_interpolation_node_3d_list);
+		}
+	}
+	_set_physics_interpolated_client_side(false);
+}
+
+Transform3D Node3D::_get_global_transform_interpolated(real_t p_interpolation_fraction) {
+	ERR_FAIL_COND_V(is_inside_tree(), Transform3D());
+
+	// Set in motion the mechanisms for client side interpolation if not already active.
+	if (!_is_physics_interpolated_client_side()) {
+		_set_physics_interpolated_client_side(true);
+
+		ERR_FAIL_NULL_V(data.client_physics_interpolation_data, Transform3D());
+		data.client_physics_interpolation_data = memnew(ClientPhysicsInterpolationData);
+		data.client_physics_interpolation_data->global_xform_curr = get_global_transform();
+		data.client_physics_interpolation_data->global_xform_prev = data.client_physics_interpolation_data->global_xform_curr;
+		data.client_physics_interpolation_data->current_physics_tick = Engine::get_singleton()->get_physics_frames();
+	}
+
+	// Storing the last tick we requested client interpolation allows us to timeout
+	// and remove client interpolated nodes from the list to save processing.
+	// We use some arbitrary timeout here, but this could potentially be user defined.
+
+	// Note: This timeout has to be larger than the number of ticks in a frame, otherwise the interpolated
+	// data will stop flowing before the next frame is drawn. This should only be relevant at high tick rates.
+	// We could alternatively do this by frames rather than ticks and avoid this problem, but then the behavior
+	// would be machine dependent.
+	data.client_physics_interpolation_data->timeout_physics_tick = Engine::get_singleton()->get_physics_frames() + 256;
+
+	// Make sure data is up to date.
+	update_client_physics_interpolation_data();
```

Comment:
This seems to be called twice per process frame - once here and other time due to calling `tree->client_physics_interpolation_add_node_3d(...)` at the end of this function. Since `update_client_physics_interpolation_data()` is not idempotent the variables `global_xform_curr` and `global_xform_curr` will end up having same value which results this function returning transforms that are not interpolated as `(a + (b - a) * f` equals `a` when `a == b`.

Reply:
Yes but see the method, it doesn't update the previous if it's on the same tick, right?

Reply:
It is often called twice, yes. The function contains a tick counter which ensures it only operates _once_ per tick.

The data must get updated, but if `get_global_transform_interpolated()` is called before the tick update, it is refreshed early. The logic admittedly isn't super clear, but it works this way to prevent an order of operations bug.

Reply:
Yes, it is true that `global_xform_prev` is updated only once per tick.

The not so obvious thing is that when `update_client_physics_interpolation_data()` gets called from `SceneTree::ClientPhysicsInterpolation::physics_process()`, the function `Engine::get_singleton()->get_physics_frames()` is returning incorrect value due to being updated late. This way the newly changed value from `get_global_transform()` is stored into `global_xform_curr` even before we actually change `global_xform_prev`. The next call to `update_client_physics_interpolation_data()` from `get_global_transform_interpolated()` we finally see that the tick has changed and we set `global_xform_prev` to `global_xform_curr`. This second call happens in the same process frame as the first one and so both `global_xform_prev` and `global_xform_curr` end up having the same value.